streamer/src/quic.rs

```diff
@@ -0,0 +1,188 @@
+        .into_iter()
+        .map(|cert| Pem {
+            tag: "CERTIFICATE".to_string(),
+            contents: cert.0,
+        })
+        .collect();
+    let cert_chain_pem = pem::encode_many(&cert_chain_pem_parts);
+
+    let mut server_config = ServerConfig::with_single_cert(cert_chain, priv_key)?;
+    let config = Arc::get_mut(&mut server_config.transport).unwrap();
+    config.max_concurrent_uni_streams(1_u8.into());
+    config.stream_receive_window(quinn::VarInt::from_u64(PACKET_DATA_SIZE as u64).unwrap());
+
+    // disable bidi & datagrams
+    config.max_concurrent_bidi_streams(0_u8.into());
+    config.datagram_receive_buffer_size(None);
+
+    Ok((server_config, cert_chain_pem))
+}
+
+fn new_cert(
+    identity_keypair: &Keypair,
+    san: IpAddr,
+) -> Result<(Vec<rustls::Certificate>, rustls::PrivateKey), Box<dyn Error>> {
+    let cert_params = new_cert_params(identity_keypair, san);
+    let cert = rcgen::Certificate::from_params(cert_params)?;
+    let cert_der = cert.serialize_der().unwrap();
+    let priv_key = cert.serialize_private_key_der();
+    let priv_key = rustls::PrivateKey(priv_key);
+    let cert_chain = vec![rustls::Certificate(cert_der)];
+    Ok((cert_chain, priv_key))
+}
+
+fn new_cert_params(identity_keypair: &Keypair, san: IpAddr) -> CertificateParams {
+    // TODO(terorie): Is it safe to sign the TLS cert with the identity private key?
+
+    // Unfortunately, rcgen does not accept a "raw" Ed25519 key.
+    // We have to convert it to DER and pass it to the library.
+
+    // Convert private key into PKCS#8 v1 object.
+    // RFC 8410, Section 7: Private Key Format
+    // https://datatracker.ietf.org/doc/html/rfc8410#section-7
+    let mut private_key = Vec::<u8>::with_capacity(34);
+    private_key.extend_from_slice(&[0x04, 0x20]); // ASN.1 OCTET STRING
+    private_key.extend_from_slice(identity_keypair.secret().as_bytes());
+    let key_pkcs8 = pkcs8::PrivateKeyInfo {
+        algorithm: AlgorithmIdentifier {
+            oid: ObjectIdentifier::from_arcs(&[1, 3, 101, 112]).unwrap(),
+            parameters: None,
+        },
+        private_key: &private_key,
+        public_key: None,
+    };
+    let key_pkcs8_der = key_pkcs8
+        .to_der()
+        .expect("Failed to convert keypair to DER")
+        .to_der();
+
+    // Parse private key into rcgen::KeyPair struct.
+    let keypair =
+        rcgen::KeyPair::from_der(&key_pkcs8_der).expect("Failed to parse keypair from DER");
+
+    let mut cert_params = CertificateParams::default();
+    cert_params.subject_alt_names = vec![SanType::IpAddress(san)];
+    cert_params.alg = &rcgen::PKCS_ED25519;
+    cert_params.key_pair = Some(keypair);
+    cert_params.distinguished_name = DistinguishedName::new();
+    cert_params
+        .distinguished_name
+        .push(DnType::CommonName, "Solana node");
+    cert_params
+}
+
+fn rt() -> Runtime {
+    Builder::new_current_thread().enable_all().build().unwrap()
+}
+
+pub fn spawn_server(
```

Comment:
Seems like this whole function is still very green. Only handles one connection at a time, just echos whatever it receives back to the sender, no error handling

Reply:
each connection should be in an async task, no?
what do you mean echos back to the sender? It sends the packet along into sigverify/tpu, similar to the streamer.